### PR TITLE
Add a custom rule for downloading the mono package.

### PR DESCRIPTION
**Pull request by zaphar**
Only works for OS X right now. Linux packages will have
to wait till later.

### CODE REVIEW

**Comment by zaphar** [comment]
First pass at this just for OS X at the moment.

**Comment by zaphar** [comment]
This is *not* ready to import since it will break linux platforms and only works on OS X. That said I'd like some early feedback before I tackle the linux version of this. Particularly on how to properly use the configuration to figure out which mono instance to use on linux or other platforms.

Should I name the repository something unique to OS X and use config values to choose the right platform specific repository for mono? Or should/can I make the repository rule figure out which mono package to pull down?

**Comment by damienmg** [comment]
I think the change is pretty good. Keeping the same name and have the repository rule figure out the correct one to download is better because a select would download both versions.

Two minor concerns:
  * the download method might increase the load to download.mono.project.
  * don't they have other archive kind that xar so you don't have to shell out? We can also mirror it on our GCS bucket if you prefer to package it in an easier format.

**Comment by zaphar** [comment]
I asked in the official mono irc channel and they said this was basically the only other option. If google is willing to mirror it then that would probably significantly improve this rule. However it adds a maintenance burden on keeping the mirror up to date.

**Comment by damienmg** [comment]
Well we can mirror it, we have to review the change anyway that's almost the same load. Do you want we mirror it? Do you want it to be repackaged in tar or zip?

**Comment by zaphar** [comment]
I'd much prefer to mirror it I think. Tar or zip is fine. I'll put together the archive with the build file and then the rule will get much simpler. Where should I send the file once it's ready?

**Comment by damienmg** [comment]
Humm let me see how we can set that up. Maybe just send it using Google drive?

**Comment by zaphar** [comment]
Okay I've put together a tarball and placed it here: https://drive.google.com/open?id=0Bzc_lt_M3d9FdW9pdDdzc3FpUE0

When you've got it in the appropriate location let me know and I'll update this change with the new rules.

**Comment by damienmg** [comment]
Sorry for taking so long, here the file is mirrored:
http://bazel-mirror.storage.googleapis.com/download.mono-project.com/archive/4.2.3/macos-10-x86/MonoFramework-MDK-4.2.3.4.macos10.xamarin.x86.tar.gz

**Comment by damienmg** [comment]
(I kept the name of the original file, just changed the extension)

**Comment by damienmg** [comment]
Ok sorry for the previous message, it is your file, renamed into the original file name with a tar gz extension

**Comment by zaphar** [comment]
No problem. This isn't my day job so I'm probably taking longer than you on turnaround.

I'm attempting to make this work and bazel is complaining that it can't download the file with the following error output using --logging=6

```____Downloading from http://bazel-mirror.storage.googleapis.com/download.mono-project.com/archive/4.2.3/macos-10-x86/MonoFramework-MDK-4.2.3.4.macos10.xamarin.x86.tar.gz: 0B
ERROR: java.io.IOException: Error downloading from http://bazel-mirror.storage.googleapis.com/download.mono-project.com/archive/4.2.3/macos-10-x86/MonoFramework-MDK-4.2.3.4.macos10.xamarin.x86.tar.gz to /private/var/tmp/_bazel_jwall/a4f05aced329c5d1ca55a18062777a02/external/mono/mono.tar.gz: Error downloading http://bazel-mirror.storage.googleapis.com/download.mono-project.com/archive/4.2.3/macos-10-x86/MonoFramework-MDK-4.2.3.4.macos10.xamarin.x86.tar.gz to /private/var/tmp/_bazel_jwall/a4f05aced329c5d1ca55a18062777a02/external/mono/mono.tar.gz/MonoFramework-MDK-4.2.3.4.macos10.xamarin.x86.tar.gz: /private/var/tmp/_bazel_jwall/a4f05aced329c5d1ca55a18062777a02/external/mono/mono.tar.gz/MonoFramework-MDK-4.2.3.4.macos10.xamarin.x86.tar.gz (No such file or directory).```

**Comment by zaphar** [comment]
Looks like my issue is caused by https://github.com/bazelbuild/bazel/issues/1235. Using the workaround there.

**Comment by zaphar** [comment]
@damienmg 

So apparently I accidentally uploaded an archive with a broken mcs script. Didn't realize it till I tried to test the download. Can you upload this archive instead? https://drive.google.com/open?id=0Bzc_lt_M3d9FZ19NTmowLTJ0bms

**Comment by damienmg** [comment]
Ok updated.

**Comment by zaphar** [comment]
I'm getting 403 access denied when trying to download.

xml response looks like this.
```xml
<Error>
<Code>AccessDenied</Code>
<Message>Access denied.</Message>
<Details>
Anonymous callers do not have storage.objects.get access to object bazel-mirror/download.mono-project.com/archive/4.2.3/macos-10-x86/MonoFramework-MDK-4.2.3.4.macos10.xamarin.x86.tar.gz.
</Details>
</Error>
```

**Comment by damienmg** [comment]
Ok sorry for that. fixed

**Comment by zaphar** [comment]
Okay updated the rules with the new binary package.

**Comment by zaphar** [comment]
Okay I think this is ready for another look.

It *should* intelligently handle things on OSX and fallback to using a locally installed mono on other platforms. It also allows you to override the download and force the use of a locally installed mono if desired.

File: README.md [hunk 1]
```diff
@@ -49,8 +47,10 @@ nuget_package(
 ```
 
 The `csharp_repositories` rule fetches external dependencies, namely the NUnit
-binaries, and the `csharp_configure` rule sets up the rules to use the local
-Mono C# toolchain installed on your system.
+binaries and the mono repository. Setting `use_local_mono` to True will use
+your installed mono framework instead of downloading one. If you are on OS X
+you can set use_local_mono to False and mono will be downloaded for you by
+bazel. Support for downloading mono on Linux and Windows is coming soon.
```
Comment (damienmg) [inline README.md]: Bazel does not work on Windows yet :)

File: dotnet/csharp.bzl [hunk 2]
```diff
@@ -474,20 +475,51 @@ csharp_autoconf = repository_rule(
...
-  """
-  csharp_autoconf(name = "local_config_csharp")
+def _mono_osx_repository_impl(repository_ctx):
+  # TODO(jwall): The below is necessary due to bug:
+  #   https://github.com/bazelbuild/bazel/issues/1235.
+  #   when that bug is fixed the below should be removed.
+  working_dir = repository_ctx.path("")
+  repository_ctx.file(repository_ctx.path("empty"))
+  if not working_dir.exists:
+    fail("working_dir %s does not exist! See https://github.com/bazelbuild/bazel/issues/1235" % working_dir)
```
Comment (damienmg) [inline dotnet/csharp.bzl]: These two lines should not be necessary

File: dotnet/csharp.bzl [hunk 3]
```diff
@@ -474,20 +475,51 @@ csharp_autoconf = repository_rule(
...
+
+  download_output = repository_ctx.path("")
+  #download the package
```
Comment (damienmg) [inline dotnet/csharp.bzl]: nit: space after #

File: dotnet/csharp.bzl [hunk 4]
```diff
@@ -474,20 +475,51 @@ csharp_autoconf = repository_rule(
...
+  repository_ctx.download_and_extract(
+    repository_ctx.attr._osx_pkg,
+    download_output,
+    repository_ctx.attr._osx_sha256,
+    "", "mono")
+
+  # now we create the build file.
+  toolchain_build = """
+package(default_visibility = ["//visibility:public"])
+exports_files(["mono", "mcs"])
```
Comment (damienmg) [inline dotnet/csharp.bzl]: nit: on Linux there is sandboxing and you might need to declare a dependency on all the files of the framework. Not needed yet so you might ignore it.

File: dotnet/csharp.bzl [hunk 5]
```diff
@@ -474,20 +475,51 @@ csharp_autoconf = repository_rule(
...
+"""
+  repository_ctx.file("bin/BUILD", toolchain_build)
+
+def _mono_repository_impl(repository_ctx):
+  print("os name: %s" % repository_ctx.os.name)
```
Comment (damienmg) [inline dotnet/csharp.bzl]: This will show a warning, remove.

File: dotnet/csharp.bzl [hunk 6]
```diff
@@ -474,20 +475,51 @@ csharp_autoconf = repository_rule(
...
+  if repository_ctx.attr.use_local:
+    _csharp_autoconf(repository_ctx)
+  elif repository_ctx.os.name == "mac os x":
```
Comment (damienmg) [inline dotnet/csharp.bzl]: find("mac") != -1 instead?

File: dotnet/csharp.bzl [hunk 7]
```diff
@@ -474,20 +475,51 @@ csharp_autoconf = repository_rule(
...
+    _mono_osx_repository_impl(repository_ctx)
+  else:
+    #if nothing else works look for a local version of mono
+    _csharp_autoconf(repository_ctx)
+
+mono_package = repository_rule(
+  implementation = _mono_repository_impl,
+  attrs = {
+    "_osx_pkg": attr.string(default="http://bazel-mirror.storage.googleapis.com/download.mono-project.com/archive/4.2.3/macos-10-x86/MonoFramework-MDK-4.2.3.4.macos10.xamarin.x86.tar.gz"),
```
Comment (damienmg) [inline dotnet/csharp.bzl]: why is it a hidden attribute, why not make it a constant if you don't want the user to override it?

File: dotnet/mono_unpack.sh [hunk 8]
```diff
@@ -0,0 +1,59 @@
+#!/bin/bash
```
Comment (damienmg) [inline dotnet/mono_unpack.sh]: Not used anymore, delete?

**Comment by damienmg** [comment]
That's really nice, I still have some comments.

File: README.md [hunk 9]
```diff
@@ -49,8 +47,10 @@ nuget_package(
 ```
 
 The `csharp_repositories` rule fetches external dependencies, namely the NUnit
-binaries, and the `csharp_configure` rule sets up the rules to use the local
-Mono C# toolchain installed on your system.
+binaries and the mono repository. Setting `use_local_mono` to True will use
+your installed mono framework instead of downloading one. If you are on OS X
+you can set use_local_mono to False and mono will be downloaded for you by
```
Comment (davidzchen) [inline README.md]: Nit: back quotes around `True`, `False`, and `use_local_mono` elsewhere in the text.

File: README.md [hunk 10]
```diff
@@ -34,12 +34,10 @@ git_repository(
 load(
     "@io_bazel_rules_dotnet//dotnet:csharp.bzl",
     "csharp_repositories",
-    "csharp_configure",
     nuget_package
 )
 
-csharp_repositories()
-csharp_configure()
+csharp_repositories(use_local_mono=True)
```
Comment (davidzchen) [inline README.md]: Nit: BUILD convention uses spaces around the `=` for keyword args.

**Comment by davidzchen** [comment]
This is awesome! Thanks for putting this together.

File: dotnet/csharp.bzl (hunk 4 above)
Comment (zaphar) [inline dotnet/csharp.bzl]: I'll cross that boat when I come to it for Linux I think. Declaring that dependency could be a bit of a challenge.

File: dotnet/csharp.bzl (hunk 7 above)
Comment (zaphar) [inline dotnet/csharp.bzl]: I wasn't clear on the expected style here. In some places I've seen it done this way and in others I've seen it as a constant. I don't have a preference myself. I'll change them to a constant instead.

File: dotnet/mono_unpack.sh (hunk 8 above)
Comment (zaphar) [inline dotnet/mono_unpack.sh]: Whoops I meant to do that and forgot.

**Comment by zaphar** [comment]
Okay I think I addressed all the comments.

**Comment by damienmg** [comment]
LGTM

**Pull request merged**